bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java

```diff
@@ -0,0 +1,401 @@
+        } catch (Exception e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.HANDLER_INITIALIZING_ERROR,
+                    "Unexpected internal error.");
+            logger.error("Failed to start http client or scheduler.");
+        }
+    }
+
+    public void stop() {
+        if (pollFuture != null) {
+            pollFuture.cancel(true);
+        }
+        try {
+            httpClient.stop();
+        } catch (Exception e) {
+            logger.error("Failed to stop http client.");
+        }
+    }
+
+    public void reset() {
+        // force re-authentication before the next action.
+        authentication = null;
+
+        // the configuration may be correct.
+        // if it was incorrect, the `authenticate()`-method will set the flag.
+        invalidConfiguration = false;
+    }
+
+    public void handleCommand(final ChannelUID channelUID, final Command command) {
+        if (authenticate()) {
+            handleCommandAuthenticated(channelUID, command);
+        }
+    }
+
+    public void setPin(final String pin) {
+        this.pin = pin;
+    }
+
+    public void setIp(final String ip) {
+        this.url = "http://" + ip + "/HNAP1/";
+    }
+
+    private SOAPElement moduleParameters(final String moduleID) throws SOAPException {
+        return soapFactory.createElement("ModuleID").addTextNode(moduleID);
+    }
+
+    public String queryStringState(final String action, final String name, final String module) {
+        try {
+            final Document document = post(action, newMessage(action, moduleParameters(module)), authentication);
+            return getElementByTagName(document, name);
+        } catch (CommunicationException e) {
+            logger.error("Failed to get quantity (communication error).");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+        } catch (SOAPException e) {
+            logger.error("Unexpected internal error.");
+            updateStatus(ThingStatus.OFFLINE);
+        }
+        return "";
+    }
+
+    public State queryOnOffState(final String action, final String name, final String module) {
+        final String state = queryStringState(action, name, module);
+        if ("true".contentEquals(state)) {
+            return OnOffType.ON;
+        } else if ("false".equals(state)) {
+            return OnOffType.OFF;
+        } else {
+            logger.error("Unexpected state: {}", state);
+        }
+        return UnDefType.UNDEF;
+    }
+
+    public State queryUnitState(final String action, final String name, final String module, final Unit<?> unit) {
+        final String s = queryStringState(action, name, module);
+        try {
+            if (s == null) {
+                return UnDefType.UNDEF;
+            } else {
+                return new QuantityType<>(Double.valueOf(s), unit);
+            }
+        } catch (NumberFormatException e) {
+            logger.error("Unexpected number format: {}", s);
```

Comment:
This should probably be a warn:
```suggestion
            logger.warn("Unexpected number format: {}", s);
```
See our documentation about logging levels:
https://www.openhab.org/docs/developer/guidelines.html#f-logging